README.md
![Build & Test](https://github.com/pnerg/jmx-runtime-json/workflows/Build%20&%20Test/badge.svg)
[![codecov](https://codecov.io/gh/pnerg/jmx-runtime-json/branch/master/graph/badge.svg?token=O8I3FS7RSI)](https://codecov.io/gh/pnerg/jmx-runtime-json)
[![Javadoc](http://javadoc-badge.appspot.com/org.dmonix/jmx-runtime-json.svg?label=javadoc)](http://javadoc-badge.appspot.com/org.dmonix/jmx-runtime-json)
[![Maven Central](https://maven-badges.herokuapp.com/maven-central/org.dmonix/jmx-runtime-json/badge.svg)](https://maven-badges.herokuapp.com/maven-central/org.dmonix/jmx-runtime-json)

JMX Runtime Information As Json
-------

This little library helps to generate a json out of a selected set of the runtime information available in the JVM.   
All of the data is accessible over various managed beans (mbeans) available in the JVM.

But accessing this data over JMX when running the application in a Docker container is difficult the least to say.  
Most applications however expose a web interface and possibly even health and or meta-information over HTTP.

The use case for this library is to convert JMX internal information into Json format that can be transported as e.g. a
HTTP response for debugging purposes or just to be logged.

# Usage

The Json is modularised into allowing for a builder pattern approach where one can choose to use one or more of the '
modules'.

* class-loading - `withClassLoadingInfo`
* garbage-collectors - `withGarbageCollectionInfo`
* memory - `withMemoryInfo`
* operating-system - `withOperatingSystemInfo`
* runtime - `withRuntimeInfo`
* thread - `withThreadInfo`

Builder pattern example:

```
JMXJsonBuilder.apply().withMemoryInfo().withRuntimeInfo();
```

The easiest way to produce a Json with all the data supported by the library is to use one of the `allInfo()` methods.  
One can choose to get an optional stack trace for every thread by supplying the depth desired to print (three in this
example).  
Not providing a depth defaults to zero, i.e. no stack trace at all.

```
String json = JMXJsonBuilder.allInfo(3).prettyPrint();
```

This will create a Json like the one below (cut in some places for brevity).

```
{
  "operating-system": {
    "name": "Mac OS X",
    "architecture": "x86_64",
    "version": "10.16",
    "available-processors": 16,
    "system-load-average": 2.81396484375
  },
  "memory": {
    "heap": {
      "init": 1073741824,
      "committed": 1073741824,
      "max": 1073741824,
      "used": 286342112,
      "pools": [
         {
          "name": "G1 Eden Space",
          "usage": {
            "init": 54525952,
            "committed": 46137344,
            "max": -1,
            "used": 4194304
          },
          "peak-usage": {
            "init": 54525952,
            "committed": 671088640,
            "max": -1,
            "used": 50331648
          }
        },
        ...
      ]   
    },
    "non-heap": {
      "init": 2555904,
      "committed": 178573312,
      "max": -1,
      "used": 164771184,
      "pools": [
       {
          "name": "CodeHeap 'non-nmethods'",
          "usage": {
            "init": 2555904,
            "committed": 2949120,
            "max": 7598080,
            "used": 2866048
          },
          "peak-usage": {
            "init": 2555904,
            "committed": 2949120,
            "max": 7598080,
            "used": 2885248
          },
         ...
      ]          
    }
  },
  "garbage-collectors": [
    {
      "name": "G1 Young Generation",
      "collection-count": 8,
      "collection-time": 66,
      "memory-pool-names": [
        "G1 Eden Space",
        "G1 Survivor Space",
        "G1 Old Gen"
      ]
    },
    ...
  ],
  "runtime": {
    "vm-name": "OpenJDK 64-Bit Server VM",
    "vm-vendor": "AdoptOpenJDK",
    "vm-version": "11.0.9+11",
    "uptime": 7448,
    "start-time": 1611145735991,
    "input-arguments": [
      "-Dfile.encoding=UTF-8",
      "-Xms1024m",
      "-Xmx1024m",
      "-Xss256k",
      "-XX:ReservedCodeCacheSize=128m"
    ],
    "classpath": [
      "/opt/sbt/bin/sbt-launch.jar"
    ]
  },
  "thread": {
    "current-thread-count": 46,
    "daemon-thread-count": 24,
    "peak-thread-count": 46,
    "thread-cpu-time-enabled": true,
    "thread-contention-monitoring-enabled": true,
    "threads": [
      {
        "name": "main",
        "id": 1,
        "blocked-count": 222,
        "blocked-time": 0,
        "waited-count": 564,
        "waited-time": 0,
        "state": "WAITING",
        "stack-trace": [
          "jdk.internal.misc.Unsafe.park(Native method)",
          "java.util.concurrent.locks.LockSupport.park(LockSupport.java:194)",
          "java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2081)"
        ]
      },
      {
        "name": "Reference Handler",
        "id": 2,
        "blocked-count": 8,
        "blocked-time": 0,
        "waited-count": 0,
        "waited-time": 0,
        "state": "RUNNABLE",
        "stack-trace": [
          "java.lang.ref.Reference.waitForReferencePendingList(Native method)",
          "java.lang.ref.Reference.processPendingReferences(Reference.java:241)",
          "java.lang.ref.Reference$ReferenceHandler.run(Reference.java:213)"
        ]
      },
     ...
    ]
  },
  "class-loading": {
    "loaded-classes": 16123,
    "total-loaded-classes": 16128,
    "unloaded-classes": 5
  }
}
```

## withClassLoadingInfo

Provides information on the total of all class loaders in the JVM.

```
JMXJsonBuilder.apply().withClassLoadingInfo();
```

## withGarbageCollectionInfo

Provides information on the garbage collectors in the JVM.

```
JMXJsonBuilder.apply().withGarbageCollectionInfo();
```

## withMemoryInfo

Provides insights into the allocated memory of the JVM, heap as well as off/non-heap (e.g thread stacks).     
Optionally one can get statistics on all memory pools.

```
JMXJsonBuilder.apply().withMemoryInfo(); //defaults to withMemoryInfo(false)
JMXJsonBuilder.apply().withMemoryInfo(true);
```

## withOperatingSystemInfo

Adds information on the operating system the JVM runs on.

```
JMXJsonBuilder.apply().withOperatingSystemInfo();
```

## withRuntimeInfo

Provides runtime information such as:

* when the JVM was started
* total uptime
* JVM vendor/version
* all input arguments
* classpath

```
JMXJsonBuilder.apply().withRuntimeInfo();
```

## withThreadInfo

As threads may be in abundance this is the most verbose part.   
Apart from providing general counters for threads it also provides insights into every thread in the system.

* name and id
* state
* blocked/waited counters
* blocked/waited time - These may be -1 depending on if the JVM is configured for measuring the information or not
* stack trace - The stack trace is optional and one can choose how many rows/lines from the stack trace to print for
  each thread (0 is default)

```
JMXJsonBuilder.apply().withThreadInfo(); //same as withThreadInfo(3)
JMXJsonBuilder.apply().withThreadInfo(0); //disables stack trace
```

**Note** The blocked/waited time measurements may not be enabled on the JVM.  
To enable invoke `ManagementFactory.getThreadMXBean().setThreadContentionMonitoringEnabled(true)`.   
This may have an impact on the performance so do it with caution.

## Formatting options

The Json printout can be in compact or pretty-printed format.  
There is a few functions basically with two different results, use the one where the name fit your application.  
All print related functions invoke the `toString(boolean)` the difference being the value of the argument.

| Function        | Result |
| ------------- |:-------------:|
| `compactPrint()`     | compact |
| `toString()`     | compact |
| `toString(false)`     | compact |
| `toString(true)`     | pretty-print |
| `prettyPrint()`     | pretty-print |

# Dockerized example

For a full example on the printout from a dockerized Java application refer to the [docker](docker/) submodule.  
It contains small application that easily be built into a Docker and executed locally.

# Dependencies

This library uses [minimal-json](https://github.com/ralfstx/minimal-json) for managing Json format.  
It is a small no-dependencies library fit for the purpose.

# Download

The libraries are plain Java (the Scala part is only unit tests), compiled for compatibility with Java 8+.   
Simply add the following dependency:

sbt

```
"org.dmonix" % "jmx-runtime-json" % [version]
```

maven

```
<dependency>
  <groupId>org.dmonix</groupId>
  <artifactId>jmx-runtime-json</artifactId>
  <version>[version]</version>
</dependency>
```

# License

The library is licensed under [Apache 2](LICENSE) 



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [1.4.0]

Added support for various printing options

### Added

* `toString(boolean)` - for choosing compact/pretty-print format with argument
* `compactPrint` - for compact printouts

### Changed

n/a

## [1.3.0]

Added support for operating system info and memory pools.

### Added

* `withGarbageCollectionInfo` - optional garbage collection information.

### Changed

Changed `allInfo()`

* added `withGarbageCollectionInfo`

## [1.2.0]

Added support for operating system info and memory pools.

### Added

* `withOperatingSystemInfo` - optional operating system information.
* `withMemoryInfo(boolean)` - optional memory pool information included in the _memory_ section

### Changed

Changed `allInfo()`

* added `withOperatingSystemInfo`
* changed to use `withMemoryInfo(true)`
* changed to use `withThreadInfo(3)`

## [1.1.1]

Fixed bug with missing file name and line number in the stack trace.

### Added

### Changed

Included the method name in the stack trace.   
Format is now _{class}.{method}({filename}:{line})_

## [1.1.0]

Added support for optional stack trace.

### Added

Added

* `withThreadInfo(int)`
* `allInfo(int)`

### Changed

The thread information now always contains a `stack-trace` array, might be zero lenght if no stack trace was requested
or is available for that thread.

## [1.0.0]

First release.

### Added

Support for:

- class-loading - `withClassLoadingInfo`
- memory - `withMemoryInfo`
- runtime - `withRuntimeInfo`
- thread - `withThreadInfo`

### Changed

lib/src/test/scala/org/dmonix/jmx/JMXJsonBuilderSpec.scala
/**
 * Copyright 2021 Peter Nerg
 *
 * <p>Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
 * except in compliance with the License. You may obtain a copy of the License at
 *
 * <p>http://www.apache.org/licenses/LICENSE-2.0
 *
 * <p>Unless required by applicable law or agreed to in writing, software distributed under the
 * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.dmonix.jmx

import com.eclipsesource.json.{Json, JsonArray, JsonObject, JsonValue}
import org.specs2.matcher.MatchResult
import org.specs2.mutable.Specification

import java.lang.management.ManagementFactory

/**
 * Tests for ''JMXJsonBuilderSpec''
 */
class JMXJsonBuilderSpec extends Specification {

  //enables measurement of blocked/waited time of threads
  ManagementFactory.getThreadMXBean().setThreadContentionMonitoringEnabled(true)
  private[this] type AssertJsonFunction = JsonObject => MatchResult[_]

  private[this] implicit class PimpedString(s: String) {
    def parseJson: JsonObject = Json.parse(s).asObject()
  }

  private[this] implicit class PimpedJsonObject(json: JsonObject) {
    def getLong(name: String) = getNonNull(name).asLong()

    def getString(name: String): String = getNonNull(name).asString()

    def getObject(name: String): JsonObject = getNonNull(name).asObject()

    def getInt(name: String): Int = getNonNull(name).asInt()

    def getArray(name: String): JsonArray = getNonNull(name).asArray()

    def getNonNull(name: String): JsonValue = {
      Option(json.get(name)) match {
        case Some(v) => v
        case None => throw new NullPointerException(s"No such attribute [$name]")
      }
    }

    def mustHaveAttribute(name: String): MatchResult[JsonValue] = {
      json.get(name) aka s"missing attribute '$name'" must not(beNull)
    }
  }

  private[this] def builderAsserts(builder: JMXJsonBuilder, assertJsonFunc: AssertJsonFunction) = {
    "must produce 'asJson' with the expected contents" >> {
      assertJsonFunc(builder.asJson())
    }
    "must produce 'compactPrint()' with the expected contents" >> {
      assertJsonFunc(builder.compactPrint().parseJson)
    }
    "must produce 'toString()' with the expected contents" >> {
      assertJsonFunc(builder.toString().parseJson)
    }
    "must produce 'toString(false)' with the expected contents" >> {
      assertJsonFunc(builder.toString(false).parseJson)
    }
    "must produce 'toString(true)' with the expected contents" >> {
      assertJsonFunc(builder.toString(true).parseJson)
    }
    "must produce 'prettyPrint' with the expected contents" >> {
      assertJsonFunc(builder.prettyPrint().parseJson)
    }
  }

  "withRuntime" >> {
    val builder = JMXJsonBuilder.apply().withRuntimeInfo()
    builderAsserts(builder, assertRuntimeContents)
  }

  "withThreadInfo" >> {
    "without thread stack info" >> {
      val builder = JMXJsonBuilder.apply().withThreadInfo()
      builderAsserts(builder, assertThreadContents(_, false))
    }
    "with thread stack info" >> {
      val builder = JMXJsonBuilder.apply().withThreadInfo(3)
      builderAsserts(builder, assertThreadContents(_, true))
    }
  }

  "withMemoryInfo" >> {
    "without pool information" >> {
      val builder = JMXJsonBuilder.apply().withMemoryInfo()
      builderAsserts(builder, assertMemoryContents(_, false))
    }
    "with pool information" >> {
      val builder = JMXJsonBuilder.apply().withMemoryInfo(true)
      builderAsserts(builder, assertMemoryContents(_, true))
    }
  }

  "withClassLoadingInfo" >> {
    val builder = JMXJsonBuilder.apply().withClassLoadingInfo()
    builderAsserts(builder, assertClassLoadingContents)
  }

  "withOperatingSystemInfo" >> {
    val builder = JMXJsonBuilder.apply().withOperatingSystemInfo()
    builderAsserts(builder, assertOperatingSystemInfo)
  }

  "withGarbageCollectionInfo" >> {
    val builder = JMXJsonBuilder.apply().withGarbageCollectionInfo()
    builderAsserts(builder, assertGarbageCollectionContents)
  }

  "allInfo" >> {
    def assertAllInfo(json: JsonObject, expectingStackInfo: Boolean = false): MatchResult[_] = {
      assertMemoryContents(json, true)
      assertRuntimeContents(json)
      assertThreadContents(json, expectingStackInfo)
      assertClassLoadingContents(json)
      assertOperatingSystemInfo(json)
      assertGarbageCollectionContents(json)
    }

    "without thread stack info" >> {
      val builder = JMXJsonBuilder.allInfo(0)
      builderAsserts(builder, assertAllInfo(_, false))
    }

    "default size thread stack info" >> {
      val builder = JMXJsonBuilder.allInfo()
      builderAsserts(builder, assertAllInfo(_, true))
    }

    "with thread stack info" >> {
      val builder = JMXJsonBuilder.allInfo(3)
      builderAsserts(builder, assertAllInfo(_, true))
    }
  }

  "StackTraceElement asString" >> {
    val declaringClass = "org.dmonix.FooClass"
    val methodName = "execute"
    val fileName = "JMXBuilderSpec.scala"
    val lineNumber = 69
    "with class, file and line number shall return full line" >> {
      JMXJsonBuilder.asString(new StackTraceElement(declaringClass, methodName, fileName, lineNumber)) === s"$declaringClass.$methodName($fileName:$lineNumber)"
    }
    "with class, file and no line number shall return string without line number" >> {
      JMXJsonBuilder.asString(new StackTraceElement(declaringClass, methodName, fileName, -1)) === s"$declaringClass.$methodName($fileName)"
    }
    "with class, no file and no line number shall return line with class name" >> {
      JMXJsonBuilder.asString(new StackTraceElement(declaringClass, methodName, null, -1)) === s"$declaringClass.$methodName(Unknown source)"
    }
    "with native method shall return line with class name" >> {
      //apparently line.nr -2 counts as native method...hmmm
      JMXJsonBuilder.asString(new StackTraceElement(declaringClass, methodName, fileName, -2)) === s"$declaringClass.$methodName(Native method)"
    }
  }

  private[this] def assertRuntimeContents(json: JsonObject): MatchResult[_] = {
    val obj = json.getObject("runtime")
    obj mustHaveAttribute "vm-vendor"
    obj mustHaveAttribute "vm-name"
    obj mustHaveAttribute "vm-version"
    obj mustHaveAttribute "uptime"
    obj mustHaveAttribute "start-time"
    obj mustHaveAttribute "input-arguments"
    obj mustHaveAttribute "classpath"
  }

  private[this] def assertThreadContents(json: JsonObject, expectingStackInfo: Boolean = false): MatchResult[_] = {
    val obj = json.getObject("thread")
    obj mustHaveAttribute "current-thread-count"
    obj mustHaveAttribute "daemon-thread-count"
    obj mustHaveAttribute "peak-thread-count"
    obj mustHaveAttribute "thread-cpu-time-enabled"
    obj mustHaveAttribute "thread-contention-monitoring-enabled"
    obj mustHaveAttribute "threads"

    //should be at least a few threads so we check the first one
    val threadsJson = obj.getArray("threads")
    threadsJson.size() === obj.getInt("current-thread-count")
    val thread = threadsJson.get(0).asObject()
    if (expectingStackInfo)
      thread mustHaveAttribute "stack-trace"
    thread mustHaveAttribute "name"
    thread mustHaveAttribute "id"
    thread mustHaveAttribute "blocked-count"
    thread mustHaveAttribute "blocked-time"
    thread mustHaveAttribute "waited-count"
    thread mustHaveAttribute "waited-time"
    thread mustHaveAttribute "state"
  }

  private[this] def assertMemoryContents(json: JsonObject, expectingPoolInfo: Boolean = false): MatchResult[_] = {
    val obj = json.getObject("memory")
    obj mustHaveAttribute "heap"
    obj mustHaveAttribute "non-heap"

    val heap = obj.getObject("heap")
    if (expectingPoolInfo)
      heap mustHaveAttribute "pools"
    heap mustHaveAttribute "init"
    heap mustHaveAttribute "committed"
    heap mustHaveAttribute "max"
    heap mustHaveAttribute "used"

    val nonheap = obj.getObject("non-heap")
    if (expectingPoolInfo)
      nonheap mustHaveAttribute "pools"
    nonheap mustHaveAttribute "init"
    nonheap mustHaveAttribute "committed"
    nonheap mustHaveAttribute "max"
    nonheap mustHaveAttribute "used"
  }

  private[this] def assertOperatingSystemInfo(json: JsonObject): MatchResult[_] = {
    val obj = json.getObject("operating-system")
    obj mustHaveAttribute "name"
    obj mustHaveAttribute "architecture"
    obj mustHaveAttribute "version"
    obj mustHaveAttribute "available-processors"
    obj mustHaveAttribute "system-load-average"
  }

  private[this] def assertClassLoadingContents(json: JsonObject): MatchResult[_] = {
    val obj = json.getObject("class-loading")
    obj mustHaveAttribute "loaded-classes"
    obj mustHaveAttribute "total-loaded-classes"
    obj mustHaveAttribute "unloaded-classes"
  }

  private[this] def assertGarbageCollectionContents(json: JsonObject): MatchResult[_] = {
    val array = json.getArray("garbage-collectors")

    //there should at least be one GC
    val obj = array.get(0).asObject()
    obj mustHaveAttribute "name"
    obj mustHaveAttribute "collection-count"
    obj mustHaveAttribute "collection-time"
    obj mustHaveAttribute "memory-pool-names"
  }

}


lib/src/main/java/org/dmonix/jmx/JMXJsonBuilder.java
/**
 * Copyright 2021 Peter Nerg
 *
 * <p>Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
 * except in compliance with the License. You may obtain a copy of the License at
 *
 * <p>http://www.apache.org/licenses/LICENSE-2.0
 *
 * <p>Unless required by applicable law or agreed to in writing, software distributed under the
 * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.dmonix.jmx;

import com.eclipsesource.json.Json;
import com.eclipsesource.json.JsonArray;
import com.eclipsesource.json.JsonObject;
import com.eclipsesource.json.WriterConfig;

import java.io.File;
import java.lang.management.*;

/**
 * Generates Json formatted runtime information extracted from JMX.
 *
 * <p>The class offers a builder pattern functionality allowing the user to generate a json with the
 * desired information.
 *
 * <p>E.g.
 *
 * <pre>
 *     JMXJsonBuilder.apply().withRuntimeInfo().prettyPrint();
 * </pre>
 *
 * or for convenience one can let the class produce everything it supports
 *
 * <pre>
 *     JMXJsonBuilder.allInfo().prettyPrint();
 * </pre>
 */
public class JMXJsonBuilder {
  private final JsonObject builderJson = Json.object();

  private JMXJsonBuilder() {}

  /**
   * Creates an empty instance of the builder.
   *
   * @return The created empty instance
   */
  public static JMXJsonBuilder apply() {
    return new JMXJsonBuilder();
  }

  /**
   * Creates an instance and adds everything supported by this class.
   *
   * <p>In practice the same as invoking {@link #allInfo(int) allInfo(3)}
   *
   * @return The created and populated instance
   * @see #allInfo(int)
   */
  public static JMXJsonBuilder allInfo() {
    return allInfo(3);
  }

  /**
   * Creates an instance and adds everything supported by this class.
   *
   * <p>The provided argument decides max depth/size of the stack-trace array on each thread.
   * Providing 0 will yield a empty stack-trace array.
   *
   * @param stackTraceDepth The max depth/size of the stack-trace array on each thread
   * @return The created and populated instance
   * @see #withClassLoadingInfo()
   * @see #withGarbageCollectionInfo()
   * @see #withMemoryInfo(boolean)
   * @see #withOperatingSystemInfo()
   * @see #withRuntimeInfo()
   * @see #withThreadInfo(int)
   * @since 1.1
   */
  public static JMXJsonBuilder allInfo(int stackTraceDepth) {
    return apply()
        .withOperatingSystemInfo()
        .withMemoryInfo(true)
        .withGarbageCollectionInfo()
        .withRuntimeInfo()
        .withThreadInfo(stackTraceDepth)
        .withClassLoadingInfo();
  }

  /**
   * Adds runtime information extracted from the 'RuntimeMXBean'
   *
   * <pre>
   *   "runtime": {
   *     "vm-name": "OpenJDK 64-Bit Server VM",
   *     "vm-vendor": "AdoptOpenJDK",
   *     "vm-version": "11.0.9+11",
   *     "uptime": 12573,
   *     "start-time": 1611129944546,
   *     "input-arguments": [
   *       "-Dfile.encoding=UTF-8",
   *       "-Xms1024m",
   *       "-Xmx1024m",
   *       "-Xss4M",
   *       "-XX:ReservedCodeCacheSize=128m"
   *     ],
   *     "classpath": [
   *       "/opt/sbt/bin/sbt-launch.jar"
   *     ]
   *   }
   * </pre>
   *
   * @return itself
   */
  public JMXJsonBuilder withRuntimeInfo() {
    RuntimeMXBean mbean = ManagementFactory.getRuntimeMXBean();

    JsonObject jo = Json.object();
    jo.add("vm-name", mbean.getVmName());
    jo.add("vm-vendor", mbean.getVmVendor());
    jo.add("vm-version", mbean.getVmVersion());
    jo.add("uptime", mbean.getUptime());
    jo.add("start-time", mbean.getStartTime());
    jo.add(
        "input-arguments", Json.array(mbean.getInputArguments().stream().toArray(String[]::new)));
    jo.add("classpath", Json.array(mbean.getClassPath().split(File.pathSeparator)));

    builderJson.add("runtime", jo);
    return this;
  }

  /**
   * Adds runtime information extracted from the 'ThreadMXBean' without stack trace information.
   *
   * <p>In practice the same as invoking {@link #withThreadInfo(int) withThreadInfo(0)}
   *
   * @return itself
   * @see #withThreadInfo(int)
   */
  public JMXJsonBuilder withThreadInfo() {
    return withThreadInfo(0);
  }

  /**
   * Adds runtime information extracted from the 'ThreadMXBean' with optional stack trace depth.
   *
   * <p>The blocked/waited time is only measured if the JVM is configured to measure thread
   * contention time.
   *
   * <p>The provided argument decides max depth/size of the stack-trace array on each thread. <br>
   * Providing 0 will yield a empty stack-trace array. <br>
   * E.g. invoking {@link #withThreadInfo(int) withThreadInfo(4)}
   *
   * <pre>
   *  "thread": {
   *     "current-thread-count": 46,
   *     "daemon-thread-count": 25,
   *     "peak-thread-count": 49,
   *     "thread-cpu-time-enabled": true,
   *     "thread-contention-monitoring-enabled": true,
   *     "threads": [
   *       {
   *         "name": "main",
   *         "id": 1,
   *         "blocked-count": 223,
   *         "blocked-time": -1,
   *         "waited-count": 665,
   *         "waited-time": -1,
   *         "state": "WAITING",
   *         "stack-trace": [
   *           "jdk.internal.misc.Unsafe(Native method)",
   *           "java.util.concurrent.locks.LockSupport(LockSupport.java:194)",
   *           "java.util.concurrent.ForkJoinPool(ForkJoinPool.java:1628)",
   *           "java.util.concurrent.ForkJoinWorkerThread(ForkJoinWorkerThread.java:183)"
   *         ]
   *       },
   *       ...
   *    ]
   * }
   * </pre>
   *
   * @param stackTraceDepth The max depth/size of the stack-trace array on each thread
   * @return itself
   * @since 1.1
   * @see ThreadMXBean#setThreadCpuTimeEnabled(boolean)
   * @see ThreadMXBean#setThreadContentionMonitoringEnabled(boolean)
   */
  public JMXJsonBuilder withThreadInfo(int stackTraceDepth) {
    ThreadMXBean mbean = ManagementFactory.getThreadMXBean();

    // create detailed info for each thread
    JsonArray threadArrayJson = Json.array();
    for (ThreadInfo ti : mbean.getThreadInfo(mbean.getAllThreadIds(), stackTraceDepth)) {

      JsonArray stackArray = Json.array();
      for (StackTraceElement ste : ti.getStackTrace()) {
        stackArray.add(asString(ste));
      }

      JsonObject to = Json.object();
      to.add("name", ti.getThreadName());
      to.add("id", ti.getThreadId());
      to.add("blocked-count", ti.getBlockedCount());
      to.add("blocked-time", ti.getBlockedTime());
      to.add("waited-count", ti.getWaitedCount());
      to.add("waited-time", ti.getWaitedTime());
      to.add("state", ti.getThreadState().name());
      // don't bother to add the stack array if it's empty
      if (!stackArray.isEmpty()) to.add("stack-trace", stackArray);

      threadArrayJson.add(to);
    }

    JsonObject jo = Json.object();
    jo.add("current-thread-count", mbean.getThreadCount());
    jo.add("daemon-thread-count", mbean.getDaemonThreadCount());
    jo.add("peak-thread-count", mbean.getPeakThreadCount());
    jo.add("thread-cpu-time-enabled", mbean.isThreadCpuTimeEnabled());
    jo.add("thread-contention-monitoring-enabled", mbean.isThreadContentionMonitoringEnabled());
    jo.add("threads", threadArrayJson);

    builderJson.add("thread", jo);
    return this;
  }

  /**
   * Adds runtime information extracted from the 'MemoryMXBean'.
   *
   * <p>In practice the same as invoking {@link #withMemoryInfo(boolean) withMemoryInfo(false)}
   *
   * <pre>
   *   "memory": {
   *     "heap": {
   *       "init": 1073741824,
   *       "committed": 1073741824,
   *       "max": 17179869184,
   *       "used": 4194304
   *     },
   *     "non-heap": {
   *       "init": 7667712,
   *       "committed": 32374784,
   *       "max": -1,
   *       "used": 28859904
   *     }
   *   }
   * </pre>
   *
   * @return itself
   * @see #withMemoryInfo(boolean)
   */
  public JMXJsonBuilder withMemoryInfo() {
    return withMemoryInfo(false);
  }

  /**
   * Adds runtime information extracted from the 'MemoryMXBean'.
   *
   * <p>The <i>pool</i> are optional.
   *
   * <pre>
   *   "memory": {
   *     "heap": {
   *       "init": 1073741824,
   *       "committed": 1073741824,
   *       "max": 17179869184,
   *       "used": 4194304,
   *       "pools": [
   *         {
   *           "name": "G1 Eden Space",
   *           "usage": {
   *             "init": 54525952,
   *             "committed": 46137344,
   *             "max": -1,
   *             "used": 4194304
   *           },
   *        ...
   *       ]
   *     },
   *     "non-heap": {
   *       "init": 7667712,
   *       "committed": 32374784,
   *       "max": -1,
   *       "used": 28859904
   *       "pools": [
   *         {
   *           "name": "CodeHeap 'non-nmethods'",
   *           "usage": {
   *             "init": 2555904,
   *             "committed": 2949120,
   *             "max": 7598080,
   *             "used": 2865920
   *           },
   *         ...
   *       ]
   *     }
   *   }
   * </pre>
   *
   * @param includeMemoryPools If details on each individual memory pool shall be provided
   * @return itself
   * @since 1.2
   */
  public JMXJsonBuilder withMemoryInfo(boolean includeMemoryPools) {
    MemoryMXBean mbean = ManagementFactory.getMemoryMXBean();

    JsonObject jo = Json.object();
    JsonObject heap = memoryAsJson(mbean.getHeapMemoryUsage());
    JsonObject nonHeap = memoryAsJson(mbean.getNonHeapMemoryUsage());

    if (includeMemoryPools) {
      JsonArray heapPools = Json.array();
      JsonArray nonHeapPools = Json.array();
      ManagementFactory.getMemoryPoolMXBeans().stream()
          .filter(MemoryPoolMXBean::isValid)
          .forEach(
              poolMBean -> {
                JsonObject poolObject = Json.object();
                poolObject.add("name", poolMBean.getName());
                poolObject.add("usage", memoryAsJson(poolMBean.getUsage()));
                poolObject.add("peak-usage", memoryAsJson(poolMBean.getPeakUsage()));

                // add the json to the correct array
                if (poolMBean.getType() == MemoryType.HEAP) heapPools.add(poolObject);
                else nonHeapPools.add(poolObject);
              });

      heap.add("pools", heapPools);
      nonHeap.add("pools", nonHeapPools);
    }

    jo.add("heap", heap);
    jo.add("non-heap", nonHeap);

    builderJson.add("memory", jo);
    return this;
  }

  /**
   * Adds information on the operating system this JVM runs on.
   *
   * <pre>
   *  "operating-system": {
   *     "name": "Mac OS X",
   *     "architecture": "x86_64",
   *     "version": "10.16",
   *     "available-processors": 16,
   *     "system-load-average": 2.7890625
   *   }
   * </pre>
   *
   * @return itself
   * @since 1.2
   */
  public JMXJsonBuilder withOperatingSystemInfo() {
    OperatingSystemMXBean mbean = ManagementFactory.getOperatingSystemMXBean();

    JsonObject jo = Json.object();
    jo.add("name", mbean.getName());
    jo.add("architecture", mbean.getArch());
    jo.add("version", mbean.getVersion());
    jo.add("available-processors", mbean.getAvailableProcessors());
    jo.add("system-load-average", mbean.getSystemLoadAverage());
    builderJson.add("operating-system", jo);

    return this;
  }

  /**
   * Adds runtime information extracted from the 'ClassLoadingMXBean'
   *
   * <pre>
   *   "class-loading": {
   *     "loaded-classes": 2527,
   *     "total-loaded-classes": 2527,
   *     "unloaded-classes": 0
   *   }
   * </pre>
   *
   * @return itself
   */
  public JMXJsonBuilder withClassLoadingInfo() {
    ClassLoadingMXBean mbean = ManagementFactory.getClassLoadingMXBean();

    JsonObject jo = Json.object();
    jo.add("loaded-classes", mbean.getLoadedClassCount());
    jo.add("total-loaded-classes", mbean.getTotalLoadedClassCount());
    jo.add("unloaded-classes", mbean.getUnloadedClassCount());

    builderJson.add("class-loading", jo);
    return this;
  }

  /**
   * Adds garbage collector information extracted from the list 'GarbageCollectorMXBean's.
   *
   * @return itself
   * @since 1.3
   */
  public JMXJsonBuilder withGarbageCollectionInfo() {

    JsonArray ja = Json.array();
    for (GarbageCollectorMXBean mbean : ManagementFactory.getGarbageCollectorMXBeans()) {
      JsonObject jo = Json.object();

      jo.add("name", mbean.getName());
      jo.add("collection-count", mbean.getCollectionCount());
      jo.add("collection-time", mbean.getCollectionTime());
      jo.add("memory-pool-names", Json.array(mbean.getMemoryPoolNames()));

      ja.add(jo);
    }

    builderJson.add("garbage-collectors", ja);
    return this;
  }

  static String asString(StackTraceElement ste) {
    String fileName = ste.getFileName();
    int lineNumber = ste.getLineNumber();
    if (ste.isNativeMethod()) {
      return String.format("%s.%s(Native method)", ste.getClassName(), ste.getMethodName());
    } else if (fileName != null && lineNumber >= 0) { // we have both file name and line number
      return String.format(
          "%s.%s(%s:%d)", ste.getClassName(), ste.getMethodName(), fileName, lineNumber);
    } else if (fileName != null) { // we only have file name
      return String.format("%s.%s(%s)", ste.getClassName(), ste.getMethodName(), fileName);
    }
    return String.format("%s.%s(Unknown source)", ste.getClassName(), ste.getMethodName());
  }

  /**
   * Converts the provided memory usage info into a Json object
   *
   * @param mem
   * @return
   */
  private static JsonObject memoryAsJson(MemoryUsage mem) {
    JsonObject json = Json.object();
    json.add("init", mem.getInit());
    json.add("committed", mem.getCommitted());
    json.add("max", mem.getMax());
    json.add("used", mem.getUsed());
    return json;
  }

  /**
   * Returns the built json object
   *
   * @return The json object
   */
  public JsonObject asJson() {
    return builderJson;
  }

  /**
   * Returns the built json compact printed.
   *
   * <p>Same as 'toString(false)'
   *
   * @return The json string
   * @see #toString(boolean)
   * @since 1.4.0
   */
  public String compactPrint() {
    return toString(true);
  }

  /**
   * Returns the built json pretty printed.
   *
   * <p>Same as 'toString(true)'
   *
   * @return The json string
   * @see #toString(boolean)
   */
  public String prettyPrint() {
    return toString(true);
  }

  /**
   * Returns the built json plain printed
   *
   * <p>Same as 'toString(false)'
   *
   * @return The json string
   * @see #toString(boolean)
   */
  @Override
  public String toString() {
    return toString(false);
  }

  /**
   * Returns the built json
   *
   * @param prettyPrint If the resulting json string shall be pretty printed or not
   * @return The json string
   * @since 1.4.0
   */
  public String toString(boolean prettyPrint) {
    return prettyPrint ? builderJson.toString(WriterConfig.PRETTY_PRINT) : builderJson.toString();
  }
}
